src/OrleansRuntime/LogViews/StorageProviderLogViewAdaptor.cs

```diff
@@ -0,0 +1,360 @@
+                    
+                    Services.Verbose("read success {0}", GlobalStateCache);
+
+                    break; // successful
+                }
+                catch (Exception e)
+                {
+                    LastPrimaryException = e;
+                }
+
+                Services.Verbose("read failed");
+
+                IncreaseBackoff(ref backoff_msec);
+            }
+
+            exit_operation("ReadAsync");
+        }
+
+
+
+        Random random = null;
+
+        /// <summary>
+        /// Increase backoff
+        /// </summary>
+        public void IncreaseBackoff(ref int backoff)
+        {
+            // after first fail do not backoff yet... keep it at zero
+            if (backoff == -1) {  
+                backoff = 0; 
+                return;
+            }
+
+            if (random == null)
+                random = new Random();
+
+            // grows exponentially up to slowpoll interval
+            if (backoff < slowpollinterval)
+                backoff = (int)((backoff + random.Next(5, 15)) * 1.5);
+
+            // during slowpoll, slightly randomize
+            if (backoff > slowpollinterval)
+                   backoff = slowpollinterval + random.Next(1, 200);
+        }
+
+        protected override async Task<int> WriteAsync()
+        {
+            enter_operation("WriteAsync");
+
+            int backoffMsec = -1;
+
+            var state = CopyTentativeState();
+            var updates = GetCurrentBatchOfUpdates();
+            bool batchsuccessfullywritten = false;
+
+            var nextglobalstate = new GrainStateWithMetaDataAndETag<TLogView>(state);
+            nextglobalstate.StateAndMetaData.WriteVector = GlobalStateCache.StateAndMetaData.WriteVector;
+            nextglobalstate.StateAndMetaData.GlobalVersion = GlobalStateCache.StateAndMetaData.GlobalVersion + updates.Length;
+            nextglobalstate.ETag = GlobalStateCache.ETag;
+
+            var writebit = nextglobalstate.StateAndMetaData.ToggleBit(Services.MyClusterId);
+
+            try
+            {
+                // for manual testing
+                //await Task.Delay(5000);
+
+                await globalStorageProvider.WriteStateAsync(grainTypeName, Services.GrainReference, nextglobalstate);
+
+                LastPrimaryException = null; // successful
+
+                batchsuccessfullywritten = true;
+
+                GlobalStateCache = nextglobalstate;
+
+                Services.Verbose("write ({0} updates) success {1}", updates.Length, GlobalStateCache);
+
+            }
+            catch (Exception e)
+            {
+                LastPrimaryException = e;
+            }
+
+            if (!batchsuccessfullywritten)
+            {
+                IncreaseBackoff(ref backoffMsec);
+
+                Services.Verbose("write apparently failed {0}", nextglobalstate);
+
+                while (true) // be stubborn until we can read what is there
```

Comment:
I have refactored the retry logic - still not using AsyncExecutor, but the code is now more readable because the retry delay is tracked in a separate class (ConnectionIssue).